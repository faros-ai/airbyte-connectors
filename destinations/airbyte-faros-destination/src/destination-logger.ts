import pQueue from '@esm2cjs/p-queue';
import crypto from 'crypto';
import {
  AirbyteLog,
  AirbyteLogger,
  AirbyteLogLevel,
  AirbyteLogLevelOrder,
  AirbyteMessage,
  AirbyteSourceLog,
  isAirbyteLog,
  shouldWriteLog,
} from 'faros-airbyte-cdk';
import fs from 'fs';
import os from 'os';
import readline from 'readline';

export class LogFiles {
  private readonly srcPath: string;
  private readonly dstPath: string;

  private readonly srcStream: fs.WriteStream;
  private readonly dstStream: fs.WriteStream;

  private readonly srcWriteQueue: pQueue = new pQueue({concurrency: 1});
  private readonly dstWriteQueue: pQueue = new pQueue({concurrency: 1});

  constructor(private readonly logger: FarosDestinationLogger) {
    const logDir = os.tmpdir();
    this.srcPath = `${logDir}/src.log`;
    this.dstPath = `${logDir}/dst.log`;
    this.srcStream = fs.createWriteStream(this.srcPath, {
      flags: 'a',
      encoding: 'utf-8',
    });
    this.dstStream = fs.createWriteStream(this.dstPath, {
      flags: 'a',
      encoding: 'utf-8',
    });
  }

  writeSourceLogs(...logs: AirbyteSourceLog[]): void {
    this.srcWriteQueue.add(() => this.writeLogs(this.srcStream, logs));
  }

  writeDestinationLogs(...logs: AirbyteSourceLog[]): void {
    this.dstWriteQueue.add(() => this.writeLogs(this.dstStream, logs));
  }

  private logError(msg: string, error: any): void {
    this.logger.localWrite(
      AirbyteLog.make(AirbyteLogLevel.WARN, `${msg}: ${JSON.stringify(error)}`)
    );
  }

  private writeLogs(
    stream: fs.WriteStream,
    logs: AirbyteSourceLog[]
  ): Promise<void> | void {
    try {
      if (stream.closed || !logs.length) {
        return;
      }
      const logString =
        logs.map((log) => JSON.stringify(log)).join('\n') + '\n';
      if (!stream.write(logString)) {
        // If the stream returns false, wait for the drain event before continuing
        return new Promise((resolve) => {
          // Update maxListeners to suppress nodejs warning
          stream.setMaxListeners(stream.getMaxListeners() + 1);
          stream.once('drain', () => {
            stream.setMaxListeners(Math.max(stream.getMaxListeners() - 1, 0));
            resolve();
          });
        });
      }
    } catch (error) {
      this.logError('Failed to save logs to disk', error);
    }
  }

  async sortedLogs(
    logger?: AirbyteLogger
  ): Promise<{content: string; hash: string} | undefined> {
    try {
      await Promise.all([
        this.srcWriteQueue.onIdle(),
        this.dstWriteQueue.onIdle(),
      ]);

      this.srcStream.end();
      this.dstStream.end();

      await Promise.all([
        new Promise((resolve) => this.srcStream.once('close', resolve)),
        new Promise((resolve) => this.dstStream.once('close', resolve)),
      ]);

      const srcLogs: AirbyteSourceLog[] = [];
      const dstLogs: AirbyteSourceLog[] = [];

      logger?.debug('Gathering sync logs for uploading to Faros');
      await Promise.all([
        readFile(this.srcPath, (line) =>
          processLogLineFromFile(line, srcLogs, logger)
        ),
        readFile(this.dstPath, (line) =>
          processLogLineFromFile(line, dstLogs, logger)
        ),
      ]);
      const allLogs = srcLogs.concat(dstLogs);
      allLogs.sort((a, b) => a.timestamp - b.timestamp);
      const content = allLogs.map((log) => JSON.stringify(log)).join('\n');
      const hash = crypto.createHash('md5').update(content).digest('base64');
      logger?.debug('Finished gathering sync logs');
      return {content, hash};
    } catch (error) {
      this.logError('Failed to gather logs for uploading to Faros', error);
      return undefined;
    }
  }
}

function processLogLineFromFile(
  line: string,
  logArray: AirbyteSourceLog[],
  logger?: AirbyteLogger
): AirbyteSourceLog {
  try {
    const log = JSON.parse(line);
    logArray.push(log);
  } catch (err: any) {
    logger?.warn(
      `Failed to parse log line: ${line}. Error: ${JSON.stringify(err)}`
    );
    return undefined;
  }
}

async function readFile(
  filePath: string,
  processLine: (line: string) => void
): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const fileStream = fs.createReadStream(filePath, 'utf-8');

    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    rl.on('line', (line) => processLine(line));
    rl.on('close', () => resolve());
    fileStream.on('error', (error) => reject(error));
  });
}

export class FarosDestinationLogger extends AirbyteLogger {
  private pendingLogs: AirbyteSourceLog[] = [];
  private _logFiles: LogFiles;

  // Default to true to ensure initial logs are saved until destination can
  // determine if it should upload logs to Faros
  private _shouldSaveLogs: boolean = true;

  constructor(level?: AirbyteLogLevel) {
    super(level);
  }

  set logFiles(logFiles: LogFiles) {
    this._logFiles = logFiles;
    if (this._logFiles && this._shouldSaveLogs) {
      this._logFiles.writeDestinationLogs(...this.pendingLogs);
      this.pendingLogs.length = 0;
    }
  }

  get shouldSaveLogs(): boolean {
    return this._shouldSaveLogs;
  }

  set shouldSaveLogs(shouldSaveLogs: boolean) {
    this._shouldSaveLogs = shouldSaveLogs;
    if (!shouldSaveLogs) {
      this.pendingLogs.length = 0;
    }
  }

  localWrite(msg: AirbyteMessage): void {
    super.write(msg);
  }

  override write(msg: AirbyteMessage): void {
    super.write(msg);

    if (
      this._shouldSaveLogs &&
      isAirbyteLog(msg) &&
      shouldWriteLog(msg, this.level)
    ) {
      const syncLog: AirbyteSourceLog = {
        timestamp: Date.now(),
        message: {
          level: AirbyteLogLevelOrder(msg.log.level),
          msg: msg.log.message,
        },
      };

      if (this._logFiles) {
        this._logFiles.writeDestinationLogs(syncLog);
      } else {
        this.pendingLogs.push(syncLog);
      }
    }
  }
}
