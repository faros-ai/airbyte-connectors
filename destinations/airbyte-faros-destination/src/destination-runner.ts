import {Command} from 'commander';
import {
  AirbyteConfig,
  AirbyteDestinationRunner,
  AirbyteLogger,
  AirbyteSpec,
} from 'faros-airbyte-cdk';

import {DestinationConfig} from './common/types';
import {Converter} from './converters/converter';
import {ConverterRegistry} from './converters/converter-registry';
import {FarosDestination} from './destination';

/** Faros Destination Runner */
export class FarosDestinationRunner extends AirbyteDestinationRunner<DestinationConfig> {
  readonly program: Command;

  /**
   * Creates a new Faros Destination Runner instance
   *
   * @param specOverride overrides the default specification with a custom one
   */
  constructor(specOverride: AirbyteSpec = undefined) {
    const logger = new AirbyteLogger();
    const destination = new FarosDestination(logger, specOverride);
    super(logger, destination);
    this.program = super.mainCommand();
  }

  /**
   * Register converters for custom streams
   */
  registerConverters(...converters: ReadonlyArray<Converter>): void {
    converters.forEach(ConverterRegistry.addConverter);
  }

  /**
   * Add an additional configuration check. Handy when using spec override.
   */
  onConfigCheck(check: (config: AirbyteConfig) => Promise<void>): void {
    (this.destination as FarosDestination).onConfigCheck = check;
  }
}
