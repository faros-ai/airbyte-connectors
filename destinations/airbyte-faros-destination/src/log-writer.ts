import crypto from 'crypto';
import {AirbyteSourceLog} from 'faros-airbyte-cdk';
import fs from 'fs';
import os from 'os';
import readline from 'readline';

export class LogWriter {
  private readonly sourceLogPath: string;
  private readonly destLogPath: string;

  private readonly sourceLogStream: fs.WriteStream;
  private readonly destLogStream: fs.WriteStream;

  constructor() {
    const logDir = os.tmpdir();
    this.sourceLogPath = `${logDir}/source.log`;
    this.destLogPath = `${logDir}/dest.log`;
    this.sourceLogStream = fs.createWriteStream(this.sourceLogPath, {
      flags: 'a',
    });
    this.destLogStream = fs.createWriteStream(this.destLogPath, {flags: 'a'});
  }

  async writeSourceLog(sourceLog: AirbyteSourceLog): Promise<void> {
    await writeDataHandlingBackpressure(
      this.sourceLogStream,
      JSON.stringify(sourceLog) + '\n'
    );
  }

  async writeDestLog(sourceLog: AirbyteSourceLog): Promise<void> {
    await writeDataHandlingBackpressure(
      this.destLogStream,
      JSON.stringify(sourceLog) + '\n'
    );
  }

  async sortedLogs(): Promise<{content: string; hash: string}> {
    this.sourceLogStream.end();
    this.destLogStream.end();

    const sourceLogs: AirbyteSourceLog[] = [];
    const destLogs: AirbyteSourceLog[] = [];

    await Promise.all([
      readFileLineByLine(this.sourceLogPath, (line) => {
        sourceLogs.push(JSON.parse(line));
      }),
      readFileLineByLine(this.destLogPath, (line) => {
        destLogs.push(JSON.parse(line));
      }),
    ]);

    const sortedLogs = sourceLogs.concat(destLogs).sort((a, b) => {
      return a.timestamp - b.timestamp;
    });
    const content = sortedLogs.map((log) => JSON.stringify(log)).join('\n');
    const hash = crypto.createHash('md5').update(content).digest('base64');
    return {content, hash};
  }
}

async function readFileLineByLine(
  filePath: string,
  processLine: (line: string) => void
): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const fileStream = fs.createReadStream(filePath);

    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity,
    });

    rl.on('line', (line) => {
      processLine(line);
    });

    rl.on('close', () => {
      console.log('Finished reading the file.');
      resolve(); // Resolve the promise when file reading is complete
    });

    fileStream.on('error', (error) => {
      console.error(
        `An error occurred while reading the file: ${error.message}`
      );
      reject(error); // Reject the promise if an error occurs
    });
  });
}

async function writeDataHandlingBackpressure(
  stream: fs.WriteStream,
  data: any
): Promise<void> {
  if (!stream.write(data)) {
    // If the stream returns false, wait for the drain event before continuing
    await waitForDrain(stream);
  }
}

function waitForDrain(stream: fs.WriteStream): Promise<void> {
  return new Promise((resolve) => {
    stream.once('drain', resolve);
  });
}
