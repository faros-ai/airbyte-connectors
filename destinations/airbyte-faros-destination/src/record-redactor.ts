import {mapValues} from 'lodash';
import {SyncRedactor} from 'redact-pii';
import {Dictionary} from 'ts-essentials';

export class RecordRedactor {
  private readonly redactor: SyncRedactor;

  constructor(
    private readonly redactCustomReplace: string | undefined = undefined,
    private readonly redactCustomRegex: ReadonlyArray<string> = []
  ) {
    const customRegexRedactors = this.redactCustomRegex.map((regex) => ({
      regexpPattern: new RegExp(regex, 'gi'),
      replaceWith: this.redactCustomReplace || 'REDACTED',
    }));
    this.redactor = new SyncRedactor({
      globalReplaceWith: this.redactCustomReplace,
      customRedactors: {
        before: [...customRegexRedactors],
      },
    });
  }

  redactRecord(
    record: Dictionary<any>,
    fieldsToRedact: ReadonlyArray<string>
  ): Dictionary<any> {
    return mapValues(record, (v, k) =>
      fieldsToRedact.includes(k) && typeof v === 'string'
        ? this.redactor.redact(v)
        : v
    );
  }
}
