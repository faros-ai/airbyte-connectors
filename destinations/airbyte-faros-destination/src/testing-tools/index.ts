export * from './cli';
export * from './testing-tools';
export * from './utils';
