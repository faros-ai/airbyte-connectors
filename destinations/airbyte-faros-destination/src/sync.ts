import {AxiosRequestConfig, AxiosResponse} from 'axios';
import {AirbyteConfig, AirbyteLogger, SyncMessage} from 'faros-airbyte-cdk';
import {
  FarosClient,
  FarosClientConfig,
  makeAxiosInstanceWithRetry,
} from 'faros-js-client';
import {DEFAULT_AXIOS_CONFIG} from 'faros-js-client/lib/client';
import {Dictionary} from 'ts-essentials';
import {VError} from 'verror';

import {LogContent} from './destination-logger';

export interface Account {
  accountId: string;
  params: Dictionary<any>;
  type: string;
  local: boolean;
}

interface AccountResponse {
  account: Account;
}

export interface AccountSync {
  syncId: string;
  logId: string;
  startedAt: Date;
  endedAt?: Date;
  status: 'error' | 'running' | 'success' | 'unknown';
  warnings?: SyncMessage[];
  errors?: SyncMessage[]; // can be non-empty and status = 'success', a.k.a. "partial success"
  metrics?: Dictionary<any>;
}

export type UpdateAccountSyncProps = Omit<
  AccountSync,
  'syncId' | 'logId' | 'startedAt'
>;

interface AccountSyncResponse {
  sync: AccountSync;
}

const DEFAULT_ACCOUNT_TYPE = 'custom';

class FarosSyncClient extends FarosClient {
  constructor(
    cfg: FarosClientConfig,
    private readonly airbyteLogger?: AirbyteLogger,
    private readonly _axiosConfig?: AxiosRequestConfig
  ) {
    super(cfg, airbyteLogger?.asPino('info'), _axiosConfig);
  }

  async createAccountSync(
    accountId: string,
    startedAt: Date,
    logId?: string
  ): Promise<AccountSync | undefined> {
    const sync = syncResult(
      await this.attemptRequest(
        this.request('PUT', `/accounts/${accountId}/syncs`, {
          startedAt: startedAt.toISOString(),
          logId,
        }),
        `Failed to create sync for account ${accountId}`
      )
    );
    if (sync) {
      this.airbyteLogger?.info(`Created sync id: ${sync?.syncId}`);
    }
    return sync;
  }

  async createLocalAccount(
    accountId: string,
    graphName: string,
    redactedConfig: AirbyteConfig,
    type?: string,
    mode?: string
  ): Promise<Account | undefined> {
    this.airbyteLogger?.debug(`Creating local account ${accountId}`);
    return accountResult(
      await this.attemptRequest<AccountResponse>(
        this.request('POST', '/accounts', {
          accountId,
          params: {...redactedConfig, graphName},
          type: type ?? DEFAULT_ACCOUNT_TYPE,
          mode,
          local: true,
        }),
        `Failed to create account ${accountId}`
      )
    );
  }

  async updateLocalAccount(
    accountId: string,
    graphName: string,
    redactedConfig: AirbyteConfig,
    type?: string,
    mode?: string
  ): Promise<Account | undefined> {
    this.airbyteLogger?.debug(`Updating local account ${accountId}`);
    return accountResult(
      await this.attemptRequest<AccountResponse>(
        this.request('PUT', `/accounts/${accountId}`, {
          params: {...redactedConfig, graphName, graphql_api: 'v2'},
          type: type ?? DEFAULT_ACCOUNT_TYPE,
          mode,
          local: true,
        }),
        `Failed to update account ${accountId}`
      )
    );
  }

  async getAccount(accountId: string): Promise<Account | undefined> {
    return accountResult(
      await this.attemptRequest(this.request('GET', `/accounts/${accountId}`))
    );
  }

  async getOrCreateAccount(
    accountId: string,
    graphName: string,
    redactedConfig: AirbyteConfig
  ) {
    const account = await this.getAccount(accountId);
    if (account) {
      return account;
    }
    return this.createLocalAccount(accountId, graphName, redactedConfig);
  }

  async updateAccountSync(
    accountId: string,
    syncId: string,
    props: UpdateAccountSyncProps
  ): Promise<AccountSync | undefined> {
    return syncResult(
      await this.attemptRequest(
        this.request('PATCH', `/accounts/${accountId}/syncs/${syncId}`, {
          ...props,
          endedAt: props.endedAt?.toISOString(),
          warnings: props.warnings?.map(cleanSyncMessage) ?? [],
          errors: props.errors?.map(cleanSyncMessage) ?? [],
          metrics: props.metrics ?? {},
        }),
        `Failed to update sync ${syncId} for account ${accountId}`
      )
    );
  }

  async getAccountSyncLogFileUrl(
    accountId: string,
    syncId: string,
    hash: string
  ): Promise<string | undefined> {
    const result: {uploadUrl: string} = await this.attemptRequest(
      this.request('POST', `/accounts/${accountId}/syncs/${syncId}/logs`, {
        hash,
      }),
      `Failed to generate log file url for sync ${syncId} for account ${accountId}`
    );
    return result?.uploadUrl;
  }

  async uploadLogs(
    accountId: string,
    syncId: string,
    {content, hash}: LogContent
  ): Promise<void> {
    const skipVerify = !hash;
    this.airbyteLogger?.debug(
      'Uploading sync logs' +
        (skipVerify ? ' (skipping hash verification)' : '')
    );
    const urlResp: {uploadUrl: string} = await this.attemptRequest(
      this.request('POST', `/accounts/${accountId}/syncs/${syncId}/logs`, {
        hash,
      }),
      `Failed to generate log file url for sync ${syncId} for account ${accountId}`
    );
    if (!urlResp?.uploadUrl) {
      return;
    }
    const api = makeAxiosInstanceWithRetry(
      this._axiosConfig ?? DEFAULT_AXIOS_CONFIG,
      this.logger
    );
    const uploadResp = await this.attemptRequest(
      api.put(urlResp.uploadUrl, content, {
        headers: {
          'content-length': content.length,
          'content-md5': hash,
          'content-type': 'text/plain; charset=UTF-8',
        },
      }),
      'Failed to upload sync logs'
    );
    if (!uploadResp && !skipVerify) {
      await this.uploadLogs(accountId, syncId, {content});
    }
    this.airbyteLogger?.debug('Finished uploading sync logs');
  }

  private attemptRequest<T>(
    f: Promise<T>,
    failureMessage?: string
  ): Promise<T | undefined> {
    return f.catch((error) => {
      if (failureMessage) {
        let message = `Error: ${error?.message ?? JSON.stringify(error)}`;
        const response = VError.info(error)?.res?.data;
        if (response) {
          message += `. Response: ${JSON.stringify(response)}`;
        }
        this.airbyteLogger?.warn(`${failureMessage}. ${message}`);
        if (isAxiosResponse(error.response)) {
          const response = error.response as AxiosResponse;
          this.airbyteLogger?.warn(
            `Error response code: ${response.status}. Body: ${response.data}`
          );
        }
      }
      return undefined;
    });
  }
}

function accountResult(response?: AccountResponse): Account | undefined {
  return response?.account;
}

function syncResult(response?: AccountSyncResponse): AccountSync | undefined {
  if (!response?.sync) {
    return undefined;
  }
  return {
    ...response.sync,
    startedAt: new Date(response.sync.startedAt),
    endedAt: response.sync.endedAt
      ? new Date(response.sync.endedAt)
      : undefined,
  };
}

type CleanedSyncMessage = Omit<SyncMessage, 'type'>;

function cleanSyncMessage(m: SyncMessage): CleanedSyncMessage {
  const {type, ...rest} = m;
  return rest;
}

function isAxiosResponse(object: any): object is AxiosResponse {
  return object && object.status && typeof object.status === 'number';
}

export default FarosSyncClient;
