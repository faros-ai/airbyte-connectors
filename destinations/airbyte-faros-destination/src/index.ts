import {Command} from 'commander';

import {FarosDestinationRunner} from './destination-runner';

/**
 * The main entry point
 *
 * You can optionally provide options `exitOverride` and `configureOutput` to
 * override default behavior of the CLI when it detects an error and where it
 * writes to respectively. See https://github.com/tj/commander.js#override-exit-and-output-handling
 * for detailed info on how these options work.
 */
export function mainCommand(options?: {
  exitOverride?: boolean;
  suppressOutput?: boolean;
}): Command {
  const destinationRunner = new FarosDestinationRunner();
  const program = destinationRunner.mainCommand();

  if (options?.exitOverride) {
    program.exitOverride();
  }
  if (options?.suppressOutput) {
    program.configureOutput({
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      writeOut: () => {},
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      writeErr: () => {},
    });
  }

  return program;
}

export {Edition, InvalidRecordStrategy} from './common/types';
export {
  Converter,
  DestinationRecord,
  DestinationModel,
  StreamContext,
  StreamName,
  StreamNameSeparator,
} from './converters/converter';
export {FarosDestinationRunner} from './destination-runner';
