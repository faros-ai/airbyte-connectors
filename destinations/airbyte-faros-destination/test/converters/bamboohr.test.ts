import {AirbyteLog, AirbyteLogLevel} from 'faros-airbyte-cdk';
import _ from 'lodash';
import {getLocal} from 'mockttp';

import {Edition, InvalidRecordStrategy} from '../../src';
import {CLI, read} from '../cli';
import {initMockttp, tempConfig, testLogger} from '../testing-tools';
import {bamboohrAllStreamsLog} from './data';

describe('bamboohr', () => {
  const logger = testLogger();
  const mockttp = getLocal({debug: false, recordTraffic: false});
  const catalogPath = 'test/resources/bamboohr/catalog.json';
  let configPath: string;
  const streamNamePrefix = 'mytestsource__bamboohr__';

  beforeEach(async () => {
    await initMockttp(mockttp);
    configPath = await tempConfig(
      mockttp.url,
      InvalidRecordStrategy.SKIP,
      Edition.CLOUD,
      {},
      {
        bamboohr: {
          bootstrap_teams_from_managers: true,
          inactive_employment_history_status: ['Terminated', 'On-Leave'],
        },
      }
    );
  });

  afterEach(async () => {
    await mockttp.stop();
  });

  test('process records from all streams', async () => {
    const cli = await CLI.runWith([
      'write',
      '--config',
      configPath,
      '--catalog',
      catalogPath,
      '--dry-run',
    ]);
    cli.stdin.end(bamboohrAllStreamsLog, 'utf8');

    const stdout = await read(cli.stdout);
    logger.debug(stdout);

    const processedByStream = {
      users: 87,
    };
    const processed = _(processedByStream)
      .toPairs()
      .map((v) => [`${streamNamePrefix}${v[0]}`, v[1]])
      .orderBy(0, 'asc')
      .fromPairs()
      .value();

    const writtenByModel = {
      geo_Address: 87,
      geo_Location: 87,
      identity_Identity: 87,
      org_Department: 9,
      org_Employee: 87,
      org_Team: 26,
      org_TeamMembership: 110,
    };

    const processedTotal = _(processedByStream).values().sum();
    const writtenTotal = _(writtenByModel).values().sum();
    expect(stdout).toMatch(`Processed ${processedTotal} records`);
    expect(stdout).toMatch(`Would write ${writtenTotal} records`);
    expect(stdout).toMatch('Errored 0 records');
    expect(stdout).toMatch(
      JSON.stringify(
        AirbyteLog.make(
          AirbyteLogLevel.INFO,
          `Processed records by stream: ${JSON.stringify(processed)}`
        )
      )
    );
    expect(stdout).toMatch(
      JSON.stringify(
        AirbyteLog.make(
          AirbyteLogLevel.INFO,
          `Would write records by model: ${JSON.stringify(writtenByModel)}`
        )
      )
    );
    expect(await read(cli.stderr)).toBe('');
    expect(await cli.wait()).toBe(0);
  });
});
