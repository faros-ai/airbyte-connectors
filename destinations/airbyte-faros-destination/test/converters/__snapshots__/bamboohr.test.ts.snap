// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bamboohr process records from all streams 1`] = `
Array [
  "Processed 87 records",
  "Processed records by stream: {\\\\\\"mytestsource__bamboohr__users\\\\\\":87}\\"},\\"type\\":\\"LOG\\"}",
  "Would write 493 records",
  "Would write records by model: {\\\\\\"geo_Address\\\\\\":87,\\\\\\"geo_Location\\\\\\":87,\\\\\\"identity_Identity\\\\\\":87,\\\\\\"org_Department\\\\\\":9,\\\\\\"org_Employee\\\\\\":87,\\\\\\"org_Team\\\\\\":26,\\\\\\"org_TeamMembership\\\\\\":110}\\"},\\"type\\":\\"LOG\\"}",
  "Skipped 0 records",
  "Errored 0 records",
]
`;
